pkg/koordlet/qosmanager/helpers/calculator.go

```diff
@@ -0,0 +1,118 @@
+limitations under the License.
+*/
+
+package helpers
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	quotav1 "k8s.io/apiserver/pkg/quota/v1"
+	"k8s.io/klog/v2"
+
+	apiext "github.com/koordinator-sh/koordinator/apis/extension"
+	slov1alpha1 "github.com/koordinator-sh/koordinator/apis/slo/v1alpha1"
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/statesinformer"
+	"github.com/koordinator-sh/koordinator/pkg/util"
+)
+
+// CalculateFilterPodsUsed calculates the sum used of filtered pods and hostApps.
+// It returns the sum used of the filtered pods, the sum used of the filtered hostApps and the system used.
+// If hostApps not passed, the hostApps part will be counted into the system used.
+//
+// e.g. To calculate non-BE used,
+//   - filterPodsUsed := sum(podUsed[i] if f_pod(pod[i]))
+//   - filterHostAppUsed := sum(hostAppUsed[i] if f_hostApp(hostApp[i]))
+//   - systemdUsed := max(nodeReserved, nodeUsed - sum(podUsed[i]) - sum(hostAppUsed[i]))
+func CalculateFilterPodsUsed(nodeUsed float64, nodeReserved float64,
+	podMetas []*statesinformer.PodMeta, podUsedMap map[string]float64,
+	hostApps []slov1alpha1.HostApplicationSpec, hostAppMetrics map[string]float64,
+	podFilterFn func(*corev1.Pod) bool,
+	hostAppFilterFn func(*slov1alpha1.HostApplicationSpec) bool) (float64, float64, float64) {
+	var podsAllUsed, podsFilterUsed, hostAppsAllUsed, hostAppsFilterUsed float64
+
+	podMetaMap := map[string]*statesinformer.PodMeta{}
+	for _, podMeta := range podMetas {
+		podMetaMap[string(podMeta.Pod.UID)] = podMeta
+	}
+
+	for podUID, podUsed := range podUsedMap {
+		podsAllUsed += podUsed
+
+		podMeta, ok := podMetaMap[podUID]
+		if !ok { // NOTE: consider podMeta-missing pods as filtered
+			klog.V(4).Infof("pod metric not included in the podMetas, uid %v", podUID)
+			podsFilterUsed += podUsed
+		} else if podFilterFn(podMeta.Pod) {
+			podsFilterUsed += podUsed
+		}
+	}
+
+	for _, hostApp := range hostApps {
+		hostAppUsed, exist := hostAppMetrics[hostApp.Name]
+		if !exist {
+			klog.V(4).Infof("host app metric not included in the hostAppMetrics, name %v", hostApp.Name)
+			continue
+		}
+
+		hostAppsAllUsed += hostAppUsed
+		if hostAppFilterFn(&hostApp) {
+			hostAppsFilterUsed += hostAppUsed
+		}
+	}
+
+	// systemUsed means the remain used excluding the filtered pods and hostApps
+	systemUsed := nodeUsed - podsAllUsed - hostAppsAllUsed
+	if systemUsed < 0 { // set systemUsed always no less than 0
+		systemUsed = 0
+	}
+	// systemUsed = max(nodeUsed - podsAllUsed - hostAppsAllUsed, nodeAnnoReserved, nodeKubeletReserved)
+	if systemUsed < nodeReserved {
+		systemUsed = nodeReserved
+	}
+
+	return podsFilterUsed, hostAppsFilterUsed, systemUsed
+}
+
+func PodHighPriorityForBatch(pod *corev1.Pod) bool {
+	priority := apiext.GetPodPriorityClassWithDefault(pod)
+	return priority != apiext.PriorityBatch && priority != apiext.PriorityFree
+}
+
+func PodHighPriorityForBE(pod *corev1.Pod) bool {
```

Comment:
BEPodFilter

Reply:
perhaps `NonBEPodFilter` is better since it filters out BE pods.